checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -142,6 +305,94 @@ private void visitAccess(ExpressionTree indexTree, ExpressionTree arrTree) {
     }
 
     @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+        ExecutableElement elt = TreeUtils.elementFromUse(node);
+        if (!PurityUtils.isSideEffectFree(atypeFactory, elt)) {
+            checkAnnotationsInClass(
+                    elt,
+                    null,
+                    node,
+                    getCurrentPath(),
+                    UpperBoundUtil.SideEffectKind.SIDE_EFFECTING_METHOD_CALL);
+        }
+        return super.visitMethodInvocation(node, p);
+    }
+
+    @Override
+    protected void commonAssignmentCheck(
+            Tree varTree, ExpressionTree valueExp, /*@CompilerMessageKey*/ String errorKey) {
+        super.commonAssignmentCheck(varTree, valueExp, errorKey);
+        if (varTree.getKind() == Kind.VARIABLE) {
+            // Not a reassignment, so nothing to check
+            return;
+        }
+        Element elt = InternalUtils.symbol(varTree);
+
+        if (elt != null
+                && (elt.getKind() == ElementKind.FIELD
+                        || elt.getKind() == ElementKind.PARAMETER
+                        || elt.getKind() == ElementKind.LOCAL_VARIABLE)
+                && !ElementUtils.isEffectivelyFinal(elt)) {
+            TypeKind typeKind = InternalUtils.typeOf(varTree).getKind();
+            Receiver field = FlowExpressions.internalReprOf(atypeFactory, (ExpressionTree) varTree);
+            SideEffectKind kind;
+            if (elt.getKind() == ElementKind.FIELD) {
+                if (typeKind == TypeKind.ARRAY) {
+                    kind = UpperBoundUtil.SideEffectKind.ARRAY_FIELD_REASSIGNMENT;
+                } else if (!typeKind.isPrimitive()) {
+                    kind = UpperBoundUtil.SideEffectKind.NON_ARRAY_FIELD_REASSIGNMENT;
+                } else {
+                    return;
+                }
+            } else if (typeKind == TypeKind.ARRAY) {
+                kind = SideEffectKind.LOCAL_VAR_REASSIGNMENT;
+            } else {
+                return;
+            }
+            checkAnnotationsInClass(elt, field, varTree, getCurrentPath(), kind);
+        }
+    }
+
+    private void checkAnnotationsInClass(
+            Element elt,
+            Receiver canonicalTargetName,
+            Tree tree,
+            TreePath path,
+            SideEffectKind sideEffectKind) {
+        List<? extends Element> enclosedElts =
+                ElementUtils.enclosingClass(elt).getEnclosedElements();
+        List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(enclosedElts);
+        for (AnnotatedTypeMirror atm : enclosedTypes) {
+            List<Receiver> rs =
+                    UpperBoundUtil.getDependentReceivers(atm.getAnnotations(), path, atypeFactory);
```

Comment:
The path is used to resolve local variables and parameters, `#1`, correctly, so it should be a path to the location of the annotation or to a place when the same scope.  This is causing the follow code:
```
import org.checkerframework.checker.index.qual.IndexFor;

public class Parameter {
    public void method(@IndexFor("#2") int a, String[] array) {
    }

    public void method2(int a, String[] array) {
        array = new String[]{""};  // an error is issued here
    }
}
```

To issue the following error:
```
Parameter.java:11: error: Invalid format string: "[reassignment.not.permitted] Reassigning the variable %s is not permitted, because the annotation %s depends on it. Reassigning %s would invalidate that annotation." args: [@LTLengthOf(value="array", offset={}) int]
        array = new String[]{""};
        ^
1 error
```
But that code shouldn't issue a warning.

Also, note that the error was not formatted properly.

Reply:
Here's another example:
```
import org.checkerframework.checker.index.qual.IndexFor;
import org.checkerframework.checker.index.qual.IndexOrLow;

public class Parameter {
    public void method(@IndexFor("#2") int a, String[] array) {
    }

    public void method2(int a, String[] b) {
        b = new String[]{""};
        int[] array = {1};
        array = new int[0]; // error here
    }
}
```

Reply:
This is still a problem.